lib/plugin.ts

```diff
@@ -138,19 +147,97 @@ export default (imports) => new WebAssembly.Instance(mod, imports).exports;`;
   // https://github.com/WebAssembly/binaryen but it's not always installed
   // everywhere or easy to install so try to gracfully handle the case where it
   // can't be found and instead just skip this step.
-  protected wasm_opt(wasm: Buffer): Buffer {
+  protected wasmOpt(wasm: Buffer): Buffer {
     const temp1 = path.join(this.cachePath, `opt-input.wasm`);
     const temp2 = path.join(this.cachePath, `opt-output.wasm`);
     fs.writeFileSync(temp1, wasm);
     try {
       execFileSync(`wasm-opt`, [`-Os`, temp1, `-o`, temp2]);
     } catch (err) {
-      if (err.code == 'ENOENT')
+      if (err.code === "ENOENT") {
         return wasm;
+      }
       throw err;
     }
     return fs.readFileSync(temp2);
   }
+
+  protected typescript(wasm: Buffer): string {
+    const parsed = describeWasm(wasm);
+
+    let imports = `export interface FunctionImports {\n`;
+    for (const imp of parsed.imports) {
+      if (imp.kind !== "Function") {
+        continue;
+      }
+      if (imp.module !== "env") {
+        continue;
+      }
+      const signature = parsed.signatures[imp.signature];
+      imports += `  ${imp.name}(`;
+      for (let j = 0; j < signature.params.length; j++) {
+        if (j > 0) {
+          imports += `, `;
+        }
+        imports += `arg${j}: ${this.wasmTypeToTypescript(signature.params[j])}`;
+      }
+      imports += `): ${this.wasmTypeToTypescript(signature.return)};\n`;
+    }
+    imports += `}\n`;
+
+    imports += `
+export interface Imports {
+  env: FunctionImports;
+}`;
+
+    let exports = `export interface Exports {\n`;
+    for (const exp of parsed.exports) {
+      if (exp.kind === "Memory") {
+        exports += `  ${exp.name}: Memory;\n`;
+        continue;
+      }
+      if (exp.kind !== "Function") {
+        continue;
+      }
+      const func = parsed.functions[exp.index - parsed.imports.length];
```

Comment:
this only works if all your imports are functions, it is `exp.index - importedFunctionCount`

Reply:
Oops indeed!